src/Language/Spectacle/Lang/Internal.hs

```diff
@@ -1,131 +1,106 @@
 {-# LANGUAGE UndecidableInstances #-}
 
--- | An internal module defining the 'Lang' machine and its instances.
+-- | The 'Lang' monad.
 --
 -- @since 0.1.0.0
 module Language.Spectacle.Lang.Internal
-  ( -- * CEK Machine
-    Lang (..),
+  ( Lang (Pure, Yield),
     send,
-    queueApp,
-    queueCompose,
-
-    -- * Internal Effects
-    NonDet (..),
+    scope,
+    Union (Op, Scoped),
   )
 where
 
 import Control.Applicative (Alternative (empty, (<|>)))
+import Control.Monad (MonadPlus (mplus, mzero), (>=>))
 import Data.Bool (bool)
 import Data.Kind (Type)
 import GHC.TypeLits (Symbol)
 
-import Data.FTCQueue
-  ( FTCQueue,
-    ViewL (ViewL),
-    viewl,
-    (|><|),
-    pattern (:<|),
-    pattern (:|>),
-  )
-import Language.Spectacle.Lang.Control
-  ( Control (Done, Next),
-    Frame (Frame),
-    Store (Store),
-    pushSyntax,
-  )
-import Language.Spectacle.Lang.Sum (Syntax, inject, type (|>))
-import Language.Spectacle.Type.Rec (Ascribe, Rec)
+import Data.Ascript (Ascribe)
+import Data.Functor.Loom (Loom)
+import qualified Data.Functor.Loom as Loom
+import Language.Spectacle.Lang.Member (Member (inject, injectS), type (:<))
+import Language.Spectacle.Lang.Op (Op)
+import Language.Spectacle.Lang.Scoped (Effect, EffectK, Scoped)
+import Language.Spectacle.Syntax.NonDet.Internal (NonDet (Choose, Empty))
 
--- -----------------------------------------------------------------------------
+-- -------------------------------------------------------------------------------------------------
 
--- | 'Lang' is a CESK machine which hosts a set of effect kinds that represent
--- spectacle syntax along with their interpreters assigning each syntactic unit
--- an operational semantics.
+-- | 'Lang' is a CEK-style interpreter for the set of effects behind Spectacles syntax and is based
+-- on Oleg's Eff monad. 'Lang' differs from Eff in it's @ctx@ parameter and its ability to support
+-- higher-order effects.
 --
--- * @sig@ - is label set of effect kinds a 'Lang' machine holds.
+-- * The type parameter @ctx@ is a type row associating variable names to their respective types.
+-- This includes plain values from the previous frame as well as primed variables in the next frame.
 --
--- * @cxt@ - is a context for typed variables in the scope of a 'Lang' machines
--- store.
---
--- * @a@ - is type of value resulting after all effects in the scope of @sig@
--- have been discharged.
+-- * The type parameter @effs@ is the set of effects a 'Lang' is capable of performing.
 --
 -- @since 0.1.0.0
-type Lang :: [Syntax] -> [Ascribe Symbol Type] -> Type -> Type
-newtype Lang sig cxt a = Lang
-  { unLang ::
-      Rec cxt ->
-      Control sig cxt a
-  }
-
--- | 'send' pushes an identity continuation frame of the syntax @m@ into
--- 'Lang'.
+type Lang :: [Ascribe Symbol Type] -> [EffectK] -> Type -> Type
+data Lang ctx effs a where
+  Pure :: a -> Lang ctx effs a
+  Yield :: Union ctx effs a -> (a -> Lang ctx effs b) -> Lang ctx effs b
```

Comment:
Just out of curiosity, is there a technical reason you're building this like a free monad, and not a freer monad? I believe a freer monad representation like this can be more efficient, because you aren't building up a big tree, and it's just as capable:

https://github.com/evanrelf/effect/blob/1a8cec17d5da12fd90d127cff11fdd247dd445fd/src/Effect/Internal/Freer.hs#L7-L13

Also removes the requirement that effects be functors, which allows you to describe them using GADTs.

Reply:
If by "freer" you mean the freer representation from Oleg's work, than Lang here is the freer version - it just uses the type-safe version of a strong sum and doesn't use an FTCQueue for composing the continuation. The impredicative encoding of the free monad is what you have there and it can sometimes be faster to work with but can also prevent GHC from making inlining decisions in some cases. In any case I don't think we should avoid optimization work until the language is more mature. 

I should also note that theres nothing requiring that effects be functors here, only that the structure used for weaving be a functor.

Reply:
I agree that we should avoid optimization work right now and focus on parity with TLA+ and correctness. Spectacle also has different requirements of the effect system than I think regular programs may have of an effect system. We've also improved the speed of the model checker (over TLA+), a lot, by virtue of eliminating states to explore using types so I think we will be okay from a performance perspective.

Reply:
I'm more than happy to replace (~>~) with compose and (:<) with Members if it helps with readability. Agda has made me very infix happy. This sort of cosmetic change is something that would only affect anyone who wanted to contribute to Spectacle in the future since the effects are only used to specify the syntax/semantics and are otherwise an unimportant detail to a user.

Reply:
Ah I guess I misunderstood re: the freer and functor stuff.

Reply:
Generally I prefer readable names over infix operators for interfaces that concern users. However, as you noted, these are only used for internal implementation and while readability does matter for maintenance I don't think they harm readability that much. So, I think it's author's choice :)